hypervisor/src/cpuidpatch.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright © 2020, Oracle and/or its affiliates.
+//
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+//
+// Copyright © 2020, Microsft  Corporation
+//
+use kvm_bindings::CpuId;
+use kvm_ioctls::*;
+use std::result;
+pub type Result<T> = result::Result<T, Error>;
+
+// CPUID feature bits
+#[cfg(target_arch = "x86_64")]
+const TSC_DEADLINE_TIMER_ECX_BIT: u8 = 24; // tsc deadline timer ecx bit.
+#[cfg(target_arch = "x86_64")]
+const HYPERVISOR_ECX_BIT: u8 = 31; // Hypervisor ecx bit.
+
+#[allow(dead_code)]
+#[derive(Copy, Clone)]
+pub enum CpuidReg {
+    EAX,
+    EBX,
+    ECX,
+    EDX,
+}
+#[derive(Debug)]
+pub enum Error {
+    /// Cannot patch the CPU ID
+    PatchCpuId(kvm_ioctls::Error),
+}
+pub struct CpuidPatch {
+    pub function: u32,
+    pub index: u32,
+    pub flags_bit: Option<u8>,
+    pub eax_bit: Option<u8>,
+    pub ebx_bit: Option<u8>,
+    pub ecx_bit: Option<u8>,
+    pub edx_bit: Option<u8>,
+}
+pub fn patch_cpuid(kvm: &Kvm) -> Result<CpuId> {
+    let mut cpuid_patches = Vec::new();
+
+    // Patch tsc deadline timer bit
+    cpuid_patches.push(CpuidPatch {
+        function: 1,
+        index: 0,
+        flags_bit: None,
+        eax_bit: None,
+        ebx_bit: None,
+        ecx_bit: Some(TSC_DEADLINE_TIMER_ECX_BIT),
+        edx_bit: None,
+    });
+
+    // Patch hypervisor bit
+    cpuid_patches.push(CpuidPatch {
+        function: 1,
+        index: 0,
+        flags_bit: None,
+        eax_bit: None,
+        ebx_bit: None,
+        ecx_bit: Some(HYPERVISOR_ECX_BIT),
+        edx_bit: None,
+    });
+
+    // Supported CPUID
+    let mut cpuid = kvm
+        .get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)
+        .map_err(Error::PatchCpuId)?;
+
```

Comment:
This part of the file is basically the only one that's really hypervisor specific, so I think it would  methomake more sense to abstract a `get_cpuid` method through the `Vcpu` trait, and keep the whole cpuid patching implementation under vmm/src/cpu.rs.
Any other reason why you moved it here?